src/snowflake/connector/file_transfer_agent.py

```diff
@@ -367,80 +373,76 @@ def _transfer_accelerate_config(self):
-                    # unknown error
-                    logger.debug(e, exc_info=True)
+            self._use_accelerate_endpoint = SnowflakeS3Util.transfer_accelerate_config(
+                client,
+                self._stage_info)
 
-            logger.debug(
-                'use_accelerate_endpoint: %s',
-                self._use_accelerate_endpoint)
-
-    def _upload_files_in_parallel(self, file_metas: List['SnowflakeFileMeta']) -> None:
+    def _upload_files_in_parallel(self, file_metas):
         """Uploads files in parallel.
 
         Args:
             file_metas: List of metadata for files to be uploaded.
         """
-        idx = 0
-        len_file_metas = len(file_metas)
-        while idx < len_file_metas:
-            end_of_idx = idx + self._parallel if \
-                idx + self._parallel <= len_file_metas else \
-                len_file_metas
-
-            logger.debug(f'uploading files idx: {idx + 1}/{end_of_idx}')
-
-            target_meta = file_metas[idx:end_of_idx]
-            while True:
-                pool = ThreadPoolExecutor(len(target_meta))
-                results: List['SnowflakeFileMeta'] = list(pool.map(
-                    SnowflakeFileTransferAgent.upload_one_file,
-                    target_meta))
-                pool.shutdown()
-
-                # need renew AWS token?
-                retry_meta: List['SnowflakeFileMeta'] = []
-                for result_meta in results:
-                    if result_meta.result_status in [
-                        ResultStatus.RENEW_TOKEN,
-                        ResultStatus.RENEW_PRESIGNED_URL
-                    ]:
-                        retry_meta.append(result_meta)
-                    else:
-                        self._results.append(result_meta)
-
-                if len(retry_meta) == 0:
-                    # no new AWS token is required
-                    break
-                if any([result_meta.result_status == ResultStatus.RENEW_TOKEN
-                        for result_meta in results]):
-                    client = self.renew_expired_client()
-                    for result_meta in retry_meta:
-                        result_meta.client = client
-                    if end_of_idx < len_file_metas:
-                        for idx0 in range(idx + self._parallel, len_file_metas):
-                            file_metas[idx0].client = client
-                if any([result_meta.result_status == ResultStatus.RENEW_PRESIGNED_URL
-                        for result_meta in results]):
+        qtask = Queue()
+        for meta in file_metas:
+            qtask.put(meta)
+        retry_round = 0
+        while not qtask.empty():
+            len_file_metas = qtask.qsize()
+            thread_number = len_file_metas if len_file_metas < self._parallel else self._parallel
+            logger.debug(f'start {retry_round} round for {qtask.qsize()} items with {thread_number} threads')
+            pool = ThreadPoolExecutor(thread_number)
```

Comment:
we are reusing the threads, only we hit refresh_presigned, then we have to modify metas and restart

Reply:
pool can not schedule new future after shutdown().  So will start a new pool for next round. And since this only occur when errors happened, won't expect this to be regular and often during one PUT so the performance should be fine.